{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ad3e4ce4",
   "metadata": {},
   "source": [
    "- http://hgdownload.soe.ucsc.edu/goldenPath/hg38/liftOver/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0c5667ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "from os import path,listdir\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import re \n",
    "import gzip\n",
    "import pygraphviz\n",
    "import networkx as nx\n",
    "from networkx.algorithms.traversal.depth_first_search import dfs_tree\n",
    "from networkx.drawing.nx_agraph import graphviz_layout\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.path import Path\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.patheffects as patheffects\n",
    "\n",
    "import gc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5a19551d",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################\n",
    "### General methods\n",
    "##################################\n",
    "\n",
    "def Call(s):\n",
    "    subprocess.call(s, shell=True, executable='/bin/bash')\n",
    "    \n",
    "    \n",
    "def bed_sort(input_file):\n",
    "    call = \"sort -k1,1 -k2,2n -o '%s' '%s'\"%(input_file,input_file)    \n",
    "    Call(call)\n",
    "    \n",
    "    \n",
    "def gtf_sort(input_file):\n",
    "    call = \"sort -k1,1 -k4,4n -o '%s' '%s'\"%(input_file,input_file)    \n",
    "    Call(call)\n",
    "        \n",
    "        \n",
    "def microparser(line):\n",
    "    if len(line.strip())==0:\n",
    "        return None\n",
    "    X = line.strip().split('\\t')        \n",
    "    r = {a.split(' ')[0]:a.split(' ')[1].replace('\"','') for a in [x.strip() for x in X[8].split(';') if len(x.strip())>0]}\n",
    "    r.update({'chr':X[0],'start':X[3],'end':X[4],'strand':X[6],'feature':X[2]})\n",
    "    return r\n",
    "\n",
    "\n",
    "def parse_meta(text):\n",
    "    assembly = re.findall('human genome \\((.*)\\),',text)[0].strip()\n",
    "    version = re.findall('version (.*) \\(',text)[0].strip()\n",
    "    ensembl = re.findall('Ensembl (.*)\\)',text)[0].strip()\n",
    "    _date = re.findall('date: (.*)',text)[0].strip()\n",
    "    return {'gencode':version,'assembly':assembly,'ensembl':ensembl,'date':_date}\n",
    "\n",
    "##################################\n",
    "### Download gencode\n",
    "##################################\n",
    "\n",
    "def download_gencode(output_raw_path,output_gene_path,output_meta_path,max_version=41):\n",
    "    for folder in [output_raw_path,output_gene_path,output_meta_path]:\n",
    "        if not path.exists(folder):\n",
    "            os.makedirs(folder)\n",
    "    call_download = \"\"\"wget \"ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_{version}/gencode.v{version}.annotation.gtf.gz\" -P {output_raw_path}\n",
    "zgrep -P \"\\\\tgene\\\\t\" {output_raw_path}/gencode.v{version}.annotation.gtf.gz | sort -k1,1 -k4,4n | gzip > {output_gene_path}/v{version}.gtf.gz\n",
    "zcat {output_raw_path}/gencode.v{version}.annotation.gtf.gz | head -n 5 > {output_meta_path}/v{version}.txt\n",
    "\"\"\"\n",
    "    versions = list(range(5,max_version+1))\n",
    "    for version in versions:\n",
    "        output_gene = path.join(output_gene_path,'v%d.gtf.gz'%version)\n",
    "        if path.exists(output_gene):\n",
    "            print('# Already processed',output_gene)\n",
    "            continue\n",
    "        call = call_download.format(version=version,output_raw_path=output_raw_path,\n",
    "                                    output_gene_path=output_gene_path,output_meta_path=output_meta_path)\n",
    "        print(call)\n",
    "        Call(call)\n",
    "        \n",
    "\n",
    "##################################\n",
    "### Create files for index\n",
    "##################################\n",
    "\n",
    "def create_equivalence_table(output_meta_path,output_ensembl_equivalence):\n",
    "    files = sorted([path.join(output_meta_path,x) for x in listdir(output_meta_path) if x.startswith('v') and x.endswith('.txt')])\n",
    "    rows = []\n",
    "    for input_file in files:\n",
    "        with open(input_file,'r') as f:\n",
    "            text = ''.join(f.readlines()).replace('\\n',' ').strip()\n",
    "        rows.append(parse_meta(text))\n",
    "    df = pd.DataFrame(rows).sort_values(by=['date'],ascending=[False])\n",
    "    df = df[['gencode','ensembl','assembly','date']]\n",
    "    df.to_csv(output_ensembl_equivalence,index=None)\n",
    "    print('# Wrote',output_ensembl_equivalence)    \n",
    "    \n",
    "    \n",
    "def create_gene_index(output_gene_path,output_gene_index,max_version=41):\n",
    "    if not path.exists(output_gene_path):\n",
    "        os.makedirs(output_gene_path)\n",
    "    files = sorted([path.join(output_gene_path,x) for x in listdir(output_gene_path) if x.endswith('.gtf.gz')])\n",
    "    fields = ['gene_id','gene_name','gene_type','hgnc_id']\n",
    "    rows = []\n",
    "    for input_file in files:\n",
    "        version = int(re.findall('v(.*)\\.gtf\\.gz',input_file)[0])\n",
    "        print(\"Reading\",input_file)\n",
    "        with gzip.open(input_file,'rt') as f:\n",
    "            for line in f:\n",
    "                e = microparser(line)\n",
    "                d = {x:e.get(x,None) for x in fields}\n",
    "                d['gencode'] = version\n",
    "                d['gene_shortid'] = d['gene_id'].split('.')[0]\n",
    "                if 'PAR_Y' in d['gene_id']:\n",
    "                    d['gene_shortid']+='_PAR_Y'\n",
    "                coord = '%s:%s-%s(%s)'%(e['chr'],e['start'],e['end'],e['strand'])\n",
    "                d['coord'] = coord\n",
    "                rows.append(d)\n",
    "    df = pd.DataFrame(rows)\n",
    "    df = df[['gene_shortid','gencode','coord']+fields]\n",
    "    df.sort_values(by=['gene_shortid','gencode'],ascending=[True,False],inplace=True)\n",
    "\n",
    "    ## Find deprecated IDs by checking where is the last version they were used.\n",
    "    deprecated = df.sort_values(by=['gene_shortid','gencode'],ascending=[True,False]).drop_duplicates(subset=['gene_shortid'],keep='first').copy()\n",
    "    deprecated = set(deprecated[deprecated['gencode']<max_version].gene_shortid)\n",
    "    df['deprecated'] = df['gene_shortid'].apply(lambda x:int(x in deprecated))\n",
    "        \n",
    "    df.to_csv(output_gene_index,index=None,compression='gzip')\n",
    "    print('# Wrote',output_gene_index)\n",
    "        \n",
    "    call = 'cat %s | gzip > %s'%(temp_all_hg38,output_bed_index)\n",
    "    print(call)\n",
    "    Call(call)\n",
    "    \n",
    "def create_graph_index(output_gene_path,chain_hg19_hg38,output_temp='temp'):\n",
    "    if not path.exists(output_temp):\n",
    "        os.makedirs(output_temp)\n",
    "    output_gene_index = path.join(output_index_path,'gene_index.csv.gz')\n",
    "    output_graph_index = path.join(output_index_path,'gene_network.pickle')\n",
    "    \n",
    "    temp_pre20_hg19 = path.join(output_temp,'pre20.hg19.bed')\n",
    "    temp_pre20_hg38 = path.join(output_temp,'pre20.hg38.bed')\n",
    "    temp_post20_hg38 = path.join(output_temp,'post20.hg38.bed')\n",
    "    temp_genes = path.join(output_temp,'genes.hg38.bed')\n",
    "    temp_all_hg38 = path.join(output_temp,'all.hg38.bed')\n",
    "    temp_o_hg19 = path.join(output_temp,\"o.hg19\")\n",
    "    temp_o_hg38 = path.join(output_temp,\"o.hg38\")\n",
    "    temp_inter = path.join(output_temp,'intersections.tsv')\n",
    "    temp_edges = path.join(output_temp,'edges.tsv')\n",
    "    \n",
    "    # Parse gene index and prepare 1) length dictionary and 2) for bed files\n",
    "    df = pd.read_csv(output_gene_index)\n",
    "    df['chr'] = df['coord'].apply(lambda x:x.split(':')[0])\n",
    "    df['start'] = df['coord'].apply(lambda x:int(x.split(':')[1].split('-')[0])-1)\n",
    "    df['end'] = df['coord'].apply(lambda x:int(x.split('-')[1].split('(')[0]))\n",
    "    df['strand'] = df['coord'].apply(lambda x:x[-2])\n",
    "    df['score']=0\n",
    "    df['id'] = df.apply(lambda x:x['gene_shortid']+'|'+str(x['gencode']),axis=1)\n",
    "    df['length'] = df['end']-df['start']\n",
    "    L = df[['id','length']].drop_duplicates().set_index('id')['length'].to_dict()\n",
    "\n",
    "\n",
    "    # Create bed files for liftOver and intersections\n",
    "    dfx = df[df['gencode']<=19][['chr','start','end','id','score','strand']].sort_values(by=['chr','start','id'])\n",
    "    dfx.to_csv(temp_pre20_hg19,sep='\\t',index=None,header=None)\n",
    "    print('Wrote',temp_pre20_hg19)\n",
    "    dfx = df[df['gencode']>19][['chr','start','end','id','score','strand']].sort_values(by=['chr','start','id'])\n",
    "    dfx.to_csv(temp_post20_hg38,sep='\\t',index=None,header=None)\n",
    "    print('Wrote',temp_post20_hg38)\n",
    "\n",
    "    # LiftOver hg19 to hg38 of gencode<=19\n",
    "    call = \"liftOver -multiple %s %s %s %s.unmapped\"%(temp_pre20_hg19,chain_hg19_hg38,temp_pre20_hg38,temp_pre20_hg38)\n",
    "    print(call)\n",
    "    Call(call)\n",
    "\n",
    "    # Combine all gene coordinates in hg38\n",
    "    call = \"cat %s %s | sort -k1,1 -k2,2n > %s \"%(temp_pre20_hg38,temp_post20_hg38,temp_all_hg38)\n",
    "    print(call)\n",
    "    Call(call)\n",
    "\n",
    "    # Intersect from gencode v5 to gencode v19 in native hg19\n",
    "    call = \"\"\"intersectBed -sorted -s -wo -a %s -b %s | \\\n",
    "    awk 'BEGIN{OFS=FS=\"\\\\t\"}{if($4 != $10) print $4\"|\"$10,  $13}' | \\\n",
    "    sort -k1,1 | \\\n",
    "    bedtools groupby -g 1 -c 2 -o sum > \\\n",
    "    %s\n",
    "    \"\"\"\n",
    "\n",
    "    call1 = call%(temp_pre20_hg19,temp_pre20_hg19,temp_o_hg19)\n",
    "    print(call1)\n",
    "    Call(call1)\n",
    "\n",
    "    # Intersect lifted over gencode to hg38 with native hg38 (20 and onwards)\n",
    "    call1 = call%(temp_all_hg38,temp_post20_hg38,temp_o_hg38)\n",
    "    print(call1)\n",
    "    Call(call1)\n",
    "\n",
    "    # Combine intersections\n",
    "    call = \"cat %s %s > %s \"%(temp_o_hg19,temp_o_hg38,temp_inter)\n",
    "    print(call)\n",
    "    Call(call)\n",
    "\n",
    "    # Prepare tsv with edges (short gene id 1, short gene id 2, overlap of g1, gencode version of g1, gencode version of g2)\n",
    "    with open(temp_inter,'r') as f, open(temp_edges,'w') as of:\n",
    "        for line in f:\n",
    "            x = line.strip().split('\\t')\n",
    "            LI = int(x[1])\n",
    "            g1,v1,g2,v2 = x[0].split('|')                \n",
    "            id1 = g1+'|'+v1\n",
    "            id2 = g2+'|'+v2\n",
    "            o1 = round(LI/L[id1],2)\n",
    "            o2 = round(LI/L[id2],2)\n",
    "            of.write('\\t'.join(map(str,(g1,g2,o1,v1,v2)))+'\\n')\n",
    "            of.write('\\t'.join(map(str,(g2,g1,o2,v2,v1)))+'\\n')\n",
    "    print(temp_edges)\n",
    "\n",
    "    # Remove duplicated edges\n",
    "    call = \"sort -u %s -o %s\"%(temp_edges,temp_edges)\n",
    "    print(call)\n",
    "    Call(call)\n",
    "\n",
    "    # Only select genes in their first and last version\n",
    "    df = pd.read_csv(output_gene_index)\n",
    "    x1 = df.sort_values(by=['gene_shortid','gencode'],ascending=[True,False]).drop_duplicates(subset=['gene_shortid'],keep='first')\n",
    "    x2 = df.sort_values(by=['gene_shortid','gencode'],ascending=[True,True]).drop_duplicates(subset=['gene_shortid'],keep='first')\n",
    "    dfg = pd.concat((x1,x2)).drop_duplicates()\n",
    "    valid_genes = set(dfg[['gene_shortid','gencode']].itertuples(index=False, name=None))\n",
    "\n",
    "    # Load edges and filter by selected gene versions\n",
    "    dfe = pd.read_csv(temp_edges,sep='\\t',header=None)\n",
    "    ### dfe = dfe[dfe[0]!=dfe[1]] #This is not desired on genes that have not been mapped to others.\n",
    "    dfe['valid'] = dfe.apply(lambda x:int(((x[0],x[3]) in valid_genes) and ((x[1],x[4]) in valid_genes)),axis=1)\n",
    "    dfe = dfe[dfe['valid']==1]\n",
    "\n",
    "    # Index edges and vertices \n",
    "    E = dfe.apply(lambda x:((x[0],x[3]),(x[1],x[4]),{'label':x[2]}),axis=1).to_list()\n",
    "    V = valid_genes\n",
    "    genes = {}\n",
    "    for x in valid_genes:\n",
    "        if x[0] not in genes:\n",
    "            genes[x[0]] = [x]\n",
    "        else:\n",
    "            genes[x[0]].append(x)\n",
    "\n",
    "    # Create and store graph\n",
    "    G = nx.DiGraph(genes=genes)\n",
    "    G.add_nodes_from(V)\n",
    "    G.add_edges_from(E)\n",
    "    nx.write_gpickle(G,output_graph_index)\n",
    "    print('# Wrote',output_graph_index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d5e0a04d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "CURRENT_VERSION = 41\n",
    "output_download_path = '/mnt/d/GENCODE'\n",
    "output_raw_path = path.join(output_download_path,'raw')\n",
    "output_gene_path = path.join(output_download_path,'genes')\n",
    "output_meta_path = path.join(output_download_path,'metadata')\n",
    "output_temp = path.join(output_download_path,'temp')\n",
    "chain_hg19_hg38 = path.abspath(path.join('chains','hg19ToHg38.over.chain.gz'))\n",
    "\n",
    "output_index_path = path.abspath('index')\n",
    "output_ensembl_equivalence = path.join(output_index_path,'ensembl_equivalence.csv')\n",
    "output_gene_index = path.join(output_index_path,'gene_index.csv.gz')\n",
    "output_graph_index = path.join(output_index_path,'gene_network.pickle')\n",
    "output_bed_index = path.join(output_index_path,'gene_coords.hg38.bed.gz')\n",
    "\n",
    "# download_gencode(output_raw_path,output_gene_path,output_meta_path,max_version=CURRENT_VERSION)\n",
    "# create_equivalence_table(output_meta_path,output_ensembl_equivalence)\n",
    "# create_gene_index(output_gene_path,output_gene_index,max_version=CURRENT_VERSION)\n",
    "# create_graph_index(output_gene_path,chain_hg19_hg38,output_temp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4ff8e7c",
   "metadata": {},
   "source": [
    "---\n",
    "# Load indices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ddbb15bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_history_network(gene_id,DFG,G,overlap_thr = 0.1,figsize=(8,8)):\n",
    "    root = gene_id\n",
    "    ## DFS to find mapping subgraph\n",
    "    GS = set()\n",
    "    for x in G.graph['genes'][gene_id]:            \n",
    "        GS |= set([x])\n",
    "        GS |= set(G.neighbors(x))\n",
    "    gs = G.subgraph(GS).copy()\n",
    "            \n",
    "#     nx.draw_networkx(gs)\n",
    "#     plt.show()\n",
    "\n",
    "    edgeattr = {x:nx.get_edge_attributes(gs,'label')[x] for x in gs.edges()}    \n",
    "    node_idx = {x:idx for idx,x in enumerate(gs.nodes())}\n",
    "    to_remove = []\n",
    "    for e in gs.edges():\n",
    "        if e[0][0]==root and e[1][0]==root:\n",
    "            continue  \n",
    "        # Remove both edges if overlap below threshold\n",
    "        if edgeattr[e]<overlap_thr:\n",
    "            to_remove.append(e)\n",
    "            to_remove.append(e[::-1])\n",
    "            continue\n",
    "        if e[0][1]>e[1][1]:\n",
    "            to_remove.append(e)\n",
    "            continue\n",
    "        if e[0][1]==e[1][1] and node_idx[e[0]]>node_idx[e[1]]:                \n",
    "            to_remove.append(e)\n",
    "            continue                      \n",
    "    gs.remove_edges_from(set(to_remove))\n",
    "    \n",
    "#     nx.draw_networkx(gs)\n",
    "#     plt.show()\n",
    "        \n",
    "#     GS = set()\n",
    "#     for x in G.graph['genes'][gene_id]:\n",
    "#         GS |= set(dfs_tree(g2,x).nodes())\n",
    "    g2 = gs.to_undirected()\n",
    "    GS = set()\n",
    "    for x in G.graph['genes'][gene_id]:            \n",
    "        GS |= set([x])\n",
    "        GS |= set(g2.neighbors(x))\n",
    "    gs.remove_nodes_from(set(gs.nodes())-GS)\n",
    "    nodelist = gs.nodes()\n",
    "    edgelist = list(gs.edges())\n",
    "    \n",
    "    node_idx = {x:idx for idx,x in enumerate(nodelist)}\n",
    "    node_labels = {x:'%s\\nv%d'%(x[0],x[1]) for x in nodelist}\n",
    "    edge_labels = {e:\"%.1f|%.1f\"%(edgeattr[(e[0],e[1])],edgeattr[(e[1],e[0])]) for e in edgelist}\n",
    "\n",
    "    ### Init figure\n",
    "    fig,ax = plt.subplots(figsize=figsize,dpi=100)\n",
    "    pos = graphviz_layout(gs, prog='neato')\n",
    "    # pos = graphviz_layout(gs, prog='twopi')\n",
    "    # pos =graphviz_layout(gs, prog='dot')\n",
    "    # pos =graphviz_layout(gs, prog='circo')\n",
    "    nx.draw_networkx_nodes(gs, pos, nodelist=[x for x in G.graph['genes'][root]],node_color='green',node_size = 500)\n",
    "    nx.draw_networkx_nodes(gs, pos, nodelist=nodelist,node_color='lightblue',node_size = 200)\n",
    "    nx.draw_networkx_labels(gs, pos, labels=node_labels, bbox={'facecolor':'white','edgecolor':'none','alpha':0.5},\n",
    "                            verticalalignment='top',font_weight='normal',font_size=8)\n",
    "\n",
    "    nx.draw_networkx_edges(gs, pos, edgelist=edgelist,edge_color='blue',arrows=True)\n",
    "    nx.draw_networkx_edge_labels(gs,pos,edge_labels=edge_labels,\n",
    "                                 font_color='blue',font_size=8,label_pos=0.7)\n",
    "\n",
    "    ax.set_frame_on(False)\n",
    "    plt.show()\n",
    "\n",
    "##########################################################\n",
    "###########\n",
    "##########################################################\n",
    "\n",
    "def plot_history_diagram(gene_id,DFG,G,overlap_thr = 0.1):\n",
    "    root = gene_id\n",
    "    ### General graphic options\n",
    "    rh = 0.8\n",
    "    rw = 0.6\n",
    "    rhm = (1-rh)/2\n",
    "    rwm = (1-rw)/2\n",
    "    predm = 0.05\n",
    "    cmap_name = 'tab20'\n",
    "    colors = sns.color_palette(cmap_name, n_colors=20)\n",
    "    codes = [Path.MOVETO]+[Path.LINETO]*5\n",
    "\n",
    "    ### DFS to find mapping subgraph\n",
    "    GS = set()\n",
    "    for x in G.graph['genes'][gene_id]:\n",
    "        GS |= set([x])\n",
    "        GS |= set(G.neighbors(x))\n",
    "    gs = G.subgraph(GS).copy()\n",
    "\n",
    "    # nx.draw_networkx(gs)\n",
    "    # plt.show()\n",
    "\n",
    "    ### Apply overlap threshold and recompute subgraph\n",
    "    edgeattr = {x:nx.get_edge_attributes(gs,'label')[x] for x in gs.edges()}\n",
    "    to_remove = []\n",
    "    for e in gs.edges():\n",
    "        # Preserve query gene\n",
    "        if e[0][0]==root and e[1][0]==root:\n",
    "            continue\n",
    "        # Remove both edges if overlap \n",
    "        if edgeattr[e]<overlap_thr:\n",
    "            to_remove.append(e)\n",
    "            to_remove.append((e[1],e[0]))\n",
    "    ## Only keep nodes accessible to query gene\n",
    "    gs.remove_edges_from(set(to_remove))\n",
    "#     GS = set()\n",
    "#     for x in G.graph['genes'][gene_id]:\n",
    "#         GS |= set(dfs_tree(gs,x).nodes())\n",
    "    g2 = gs.to_undirected()\n",
    "    GS = set()\n",
    "    for x in G.graph['genes'][gene_id]:            \n",
    "        GS |= set([x])\n",
    "        GS |= set(g2.neighbors(x))    \n",
    "    gs.remove_nodes_from(set(gs.nodes())-GS)     \n",
    "    \n",
    "    nodelist = list(gs.nodes())\n",
    "    \n",
    "    ### Compute latest version of gene\n",
    "    edgeattr = {x:nx.get_edge_attributes(gs,'label')[x] for x in gs.edges()}\n",
    "    ### Get gene with more mutual intersection using harmonic mean\n",
    "    score = {}    \n",
    "    for e,v in edgeattr.items():\n",
    "        if e[0][0]==root or e[1][0]==root:\n",
    "            v = e[0][0] if e[1][0]==root else e[1][0]\n",
    "            _s = ((edgeattr[e]**-1+edgeattr[e[::-1]]**-1)/2)**-1\n",
    "            if v not in score:\n",
    "                score[v] = _s\n",
    "            else:\n",
    "                score[v] = max(score[v],_s)    \n",
    "                \n",
    "    x = pd.DataFrame(nodelist)\n",
    "    latest = set(x.sort_values(by=[0,1],ascending=[True,False]).drop_duplicates(subset=0,keep='first').itertuples(index=None,name=None))\n",
    "\n",
    "    ### Remove redundant edges\n",
    "    to_keep=[]\n",
    "    for e in gs.edges():\n",
    "        ## Edge must go from older to newer version\n",
    "        if e[0][1] > e[1][1]:\n",
    "            continue\n",
    "        ## Must go from oldest to latest in same gene, or connect latest to latest in different genes.\n",
    "        #if (e[0][0]==e[1][0]) or ((e[0] in latest) and (e[1] in latest)):\n",
    "        if (e[0][0]==e[1][0]) or (e[0] in latest):\n",
    "            to_keep.append(e)\n",
    "    ### Reduce graph \n",
    "    gs.remove_edges_from(set(gs.edges())-set(to_keep))\n",
    "    orphans = [x[0] for x in gs.degree() if x[1]==0]\n",
    "    gs.remove_nodes_from(set(orphans))\n",
    "    nodelist = list(gs.nodes())\n",
    "    edgelist = list(gs.edges())\n",
    "\n",
    "    ### Assign X-Y coords to genes and versions\n",
    "    _genes = []\n",
    "    _versions = []\n",
    "    for n in gs.nodes():\n",
    "        _genes.append(n[0])\n",
    "        _versions.append(n[1])\n",
    "    _genes = [root] + sorted(set(_genes)-set([root]),key=lambda x:-score[x])\n",
    "    _versions = sorted(set(_versions))\n",
    "    _genes_idx = {x:idx for idx,x in enumerate(_genes)}\n",
    "    _versions_idx = {x:idx for idx,x in enumerate(_versions)}\n",
    "    _genes_color = {g:colors[(idx*2)%20] for idx,g in enumerate(_genes)}\n",
    "\n",
    "    ### Manual position\n",
    "    pos = {}\n",
    "    for x in gs.nodes():\n",
    "        pos[x] = (_versions_idx[x[1]],_genes_idx[x[0]])\n",
    "\n",
    "    ### Compute successors and predecessors for edges' endpoints \n",
    "    succ = {}\n",
    "    pred = {}\n",
    "    for g,c in pos.items():\n",
    "        s = list(gs.successors(g))\n",
    "        ns = len(s)    \n",
    "        if ns > 0:\n",
    "            if ns == 1:\n",
    "                succ_x = {s[0]:0.5}\n",
    "            else:\n",
    "                succ_x = {z:x for z,x in zip(s,np.linspace(rhm+predm,rhm+rh-predm,len(s)))}\n",
    "            succ[g] = succ_x    \n",
    "        p = list(gs.predecessors(g))    \n",
    "        n = len(p)    \n",
    "        if n > 0:\n",
    "            if n == 1:\n",
    "                pred_x = {p[0]:0.5}\n",
    "            else:\n",
    "                pred_x = {z:x for z,x in zip(p,np.linspace(rhm+predm,rhm+rh-predm,len(p)))}\n",
    "            pred[g] = pred_x\n",
    "\n",
    "    ### Get gene info\n",
    "    df = pd.DataFrame(nodelist,columns=['gene_shortid','gencode'])\n",
    "    df = pd.merge(df,DFG).set_index(['gene_shortid','gencode'])\n",
    "    df['gversion'] = df['gene_id'].apply(lambda x:\"-\" if \".\" not in x else \"v\"+x.split('.')[1])\n",
    "    df['nodelabel'] = df.apply(lambda x:\"%s\\n%s\\n%s\"%(x['gene_name'],x['gene_type'].replace('_',' '),x['gversion']),axis=1)        \n",
    "\n",
    "    ### Init figure\n",
    "    fig,ax = plt.subplots(figsize=(2+len(_versions)*1.8,1+len(_genes)*.8),dpi=100)\n",
    "    _patches = []\n",
    "    for g,c in pos.items():\n",
    "        color = _genes_color[g[0]]\n",
    "        # Draw gene box\n",
    "        r = patches.Rectangle((_versions_idx[g[1]]+rwm,_genes_idx[g[0]]+rhm),rw,rh,\n",
    "                              linewidth=1,edgecolor=color,facecolor='none',zorder=2,alpha=1)\n",
    "        _patches.append(r)\n",
    "        # Draw gene info\n",
    "        ax.text(_versions_idx[g[1]]+.5,_genes_idx[g[0]]+.5,df.loc[g]['nodelabel'],\n",
    "                ha='center',va='center',fontsize=6,zorder=5)    \n",
    "        if g not in succ:\n",
    "            continue    \n",
    "        for s in succ[g]:\n",
    "            # Draw edge info\n",
    "            edgelabel = \"%.1f|%.1f\"%(G.edges[(g,s)]['label'],G.edges[(s,g)]['label'])\n",
    "            weight = 'bold' if edgelabel == '1.0|1.0' else 'normal'\n",
    "            ax.text(_versions_idx[s[1]]+rwm,_genes_idx[s[0]]+pred[s][g]+0.02,edgelabel,\n",
    "                    ha='right',va='top',color='k',fontsize=6,zorder=5,weight=weight)\n",
    "            zorder = 3 if g[0]==root or s[0]==root else 2\n",
    "\n",
    "            # Draw edge\n",
    "            verts = [\n",
    "                (_versions_idx[g[1]]+rwm+rw,_genes_idx[g[0]]+0.5),\n",
    "                (_versions_idx[g[1]]+1,_genes_idx[g[0]]+0.5)\n",
    "            ]\n",
    "            if (_versions_idx[s[1]] - _versions_idx[g[1]]) == 1:\n",
    "                verts.append((verts[-1][0],_genes_idx[s[0]]+pred[s][g]))\n",
    "                verts.append((verts[-1][0]+rwm,verts[-1][1]))\n",
    "                _path = Path(verts, codes[:4])\n",
    "                _patches.append(patches.PathPatch(_path, edgecolor=color, facecolor='none', \n",
    "                                                  lw=1.5,zorder=zorder,alpha=0.9))\n",
    "                continue            \n",
    "            verts.append((verts[-1][0],_genes_idx[g[0]]+1))\n",
    "            if s[1]==g[1]:\n",
    "                verts.append((_versions_idx[s[1]]+predm,_genes_idx[g[0]]+1))\n",
    "            else:\n",
    "                verts.append((_versions_idx[s[1]],_genes_idx[g[0]]+1))\n",
    "            verts.append((verts[-1][0],_genes_idx[s[0]]+pred[s][g]))\n",
    "            verts.append((_versions_idx[s[1]]+rwm,verts[-1][1]))\n",
    "            _path = Path(verts, codes)\n",
    "            _patches.append(patches.PathPatch(_path, edgecolor=color, facecolor='none', \n",
    "                                              lw=1.5,zorder=zorder,alpha=0.9))\n",
    "\n",
    "    for p in _patches:\n",
    "        ax.add_patch(p)\n",
    "\n",
    "    for y in np.arange(len(_genes)+1):\n",
    "        ax.axhline(y=y,color='gray',lw=1,zorder=1,ls='--')\n",
    "\n",
    "    ax.set_xticks(np.arange(len(_versions))+.5,_versions)    \n",
    "    ax.set_yticks(np.arange(len(_genes))+.5,_genes)\n",
    "    ax.get_yticklabels()[0].set_weight('bold')\n",
    "    ax.set_xlim(-.2,len(_versions)+.1)\n",
    "    ax.set_ylim(len(_genes)+.1,0)\n",
    "    ax.set_xlabel('Gencode version')\n",
    "    ax.set_ylabel('Stable gene id')\n",
    "\n",
    "    ax.set_frame_on(False)\n",
    "    ax.xaxis.tick_top()\n",
    "    ax.xaxis.set_label_position('top') \n",
    "    ax.tick_params(axis=u'both', which=u'both',length=0)\n",
    "    plt.show()\n",
    "    \n",
    "def plot_tracks(gene_id,DFG,G,DFC,overlap_thr = 0.1):\n",
    "    root = gene_id\n",
    "    ### General graphic options\n",
    "    rh = 0.3\n",
    "    Lm = .1    \n",
    "    cmap_name = 'tab20'\n",
    "    colors = sns.color_palette(cmap_name, n_colors=20)\n",
    "    codes = [Path.MOVETO]+[Path.LINETO]*5\n",
    "\n",
    "    ### DFS to find mapping subgraph\n",
    "    GS = set()\n",
    "    for x in G.graph['genes'][gene_id]:\n",
    "        GS |= set([x])\n",
    "        GS |= set(G.neighbors(x))\n",
    "    gs = G.subgraph(GS).copy()\n",
    "\n",
    "    ### Apply overlap threshold and recompute subgraph\n",
    "    edgeattr = {x:nx.get_edge_attributes(gs,'label')[x] for x in gs.edges()}\n",
    "    to_remove = []\n",
    "    for e in gs.edges():        \n",
    "        # Preserve query gene\n",
    "        if e[0][0]==root and e[1][0]==root:\n",
    "            continue\n",
    "        # Remove both edges if overlap \n",
    "        if edgeattr[e]<overlap_thr:\n",
    "            to_remove.append(e)\n",
    "            to_remove.append((e[1],e[0]))\n",
    "    ## Only keep nodes accessible to query gene\n",
    "    gs.remove_edges_from(set(to_remove))    \n",
    "#     GS = set()\n",
    "#     for x in G.graph['genes'][gene_id]:\n",
    "#         GS |= set(dfs_tree(gs,x).nodes())\n",
    "    g2 = gs.to_undirected()\n",
    "    GS = set()\n",
    "    for x in G.graph['genes'][gene_id]:\n",
    "        GS |= set([x])\n",
    "        GS |= set(g2.neighbors(x))    \n",
    "    gs.remove_nodes_from(set(gs.nodes())-GS)\n",
    "    nodelist = list(gs.nodes())\n",
    "    \n",
    "    ### Compute latest version of gene\n",
    "    edgeattr = {x:nx.get_edge_attributes(gs,'label')[x] for x in gs.edges()}\n",
    "    ### Get gene with more mutual intersection using harmonic mean\n",
    "    score = {}    \n",
    "    for e,v in edgeattr.items():\n",
    "        if e[0][0]==root or e[1][0]==root:\n",
    "            v = e[0][0] if e[1][0]==root else e[1][0]\n",
    "            _s = ((edgeattr[e]**-1+edgeattr[e[::-1]]**-1)/2)**-1\n",
    "            if v not in score:\n",
    "                score[v] = _s\n",
    "            else:\n",
    "                score[v] = max(score[v],_s)        \n",
    "\n",
    "    ### Assign X-Y coords to genes and versions\n",
    "    _genes = []\n",
    "    _versions = []\n",
    "    for n in gs.nodes():\n",
    "        _genes.append(n[0])\n",
    "        _versions.append(n[1])\n",
    "    _genes = [root] + sorted(set(_genes)-set([root]),key=lambda x:-score[x])\n",
    "    _versions = sorted(set(_versions))\n",
    "    _genes_idx = {x:idx for idx,x in enumerate(_genes)}\n",
    "    _versions_idx = {x:idx for idx,x in enumerate(_versions)}\n",
    "    _genes_color = {g:colors[(idx*2)%20] for idx,g in enumerate(_genes)}\n",
    "    nodelist2 = [\"%s|%s\"%x for x in nodelist]    \n",
    "    \n",
    "    ### Get gene info\n",
    "    df = pd.DataFrame(nodelist,columns=['gene_shortid','gencode'])\n",
    "    df = pd.merge(df,DFG).set_index(['gene_shortid','gencode'])\n",
    "    df['gversion'] = df['gene_id'].apply(lambda x:\"-\" if \".\" not in x else \"v\"+x.split('.')[1])\n",
    "    df['nodelabel'] = df.apply(lambda x:\"%s | %s\"%(x['gene_name'],x['gene_type'].replace('_',' ')),axis=1)\n",
    "    \n",
    "    ### Coordinates \n",
    "    dfc = DFC[DFC[3].isin(set(nodelist2))].copy()\n",
    "    dfc.rename(columns={1:'start',2:'end',5:'strand'},inplace=True)    \n",
    "    dfc[['gene_id','version']]= dfc[3].str.split('|',expand=True)\n",
    "    dfc['gene_id'] = pd.Categorical(dfc['gene_id'],categories=_genes,ordered=True)\n",
    "    dfc['version'] = dfc['version'].astype(int)\n",
    "    dfc.sort_values(by=['gene_id','version'],ascending=[True,False],inplace=True)\n",
    "    zz = dfc[['gene_id','version']].drop_duplicates().copy()\n",
    "    zz['y'] = np.arange(len(zz))\n",
    "    gene_y = zz.groupby('gene_id')['y'].mean()\n",
    "    dfc = pd.merge(dfc,zz,left_on=['gene_id','version'],right_on=['gene_id','version'])\n",
    "    nrows = dfc[['gene_id','version']].drop_duplicates().groupby('gene_id').size().to_dict()\n",
    "    \n",
    "    L = dfc.end.max()-dfc.start.min() \n",
    "    Lmed = (dfc.end.max()+dfc.start.min())/2\n",
    "    \n",
    "    #######################\n",
    "    # Plot tracks\n",
    "    #######################\n",
    "    fig,(ax0,ax) = plt.subplots(figsize=(8,len(zz)*.5+1),ncols=2,gridspec_kw={'width_ratios': [.2,3],'wspace':0})\n",
    "    _patches = []    \n",
    "    rhm = (1-rh)/2\n",
    "    for gene_id,g in dfc.groupby('gene_id',sort=False):\n",
    "        ax0.text(0,gene_y[gene_id]+.5,gene_id,ha='right',va='center',weight=('bold' if gene_id==root else 'normal'))\n",
    "        color = _genes_color[gene_id]\n",
    "        for version,g2 in g.groupby('version',sort=False): \n",
    "            y = g2.iloc[0]['y']\n",
    "            ax0.text(0.1,y+.5,version,ha='left',va='center')\n",
    "            # Draw gene info\n",
    "            ax.text(Lmed,y+.5-rh/2,df.loc[(gene_id,version)]['nodelabel'],\n",
    "                    ha='center',va='bottom',fontsize=6,zorder=5,color='black')\n",
    "            for x in g2.itertuples():                \n",
    "                r = patches.Rectangle((x.start,x.y+rhm),x.end-x.start+1,rh,\n",
    "                                      linewidth=1,facecolor=color,edgecolor='black',zorder=2,alpha=1)\n",
    "                _patches.append(r)\n",
    "    for p in _patches:\n",
    "        ax.add_patch(p)\n",
    "    for y in dfc.groupby('gene_id').y.min():\n",
    "        ax.axhline(y=y,color='gray',lw=1,zorder=1,ls='--')\n",
    "    ax.axhline(y=dfc.y.max()+1,color='gray',lw=1,zorder=1,ls='--')\n",
    "    \n",
    "    ax0.axis('off')\n",
    "    ax0.set_ylim(dfc.y.max()+1.5,-.5)\n",
    "    ax0.set_xlim(0,1)\n",
    "    ax.set_xlabel(\"%s (%s)\"%(dfc.iloc[0][0],dfc.iloc[0]['strand']))\n",
    "    ax.set_ylim(dfc.y.max()+1.5,-.5)\n",
    "           \n",
    "    ax.set_xlim(dfc.start.min()-int(L*Lm),dfc.end.max()+int(L*Lm))\n",
    "    ax.set_frame_on(False)\n",
    "    ax.xaxis.tick_top()\n",
    "    ax.xaxis.set_label_position('top') \n",
    "    ax.tick_params(axis=u'both', which=u'both',length=0)\n",
    "    ax.get_yaxis().set_ticks([])\n",
    "    plt.show()\n",
    "#     return dfc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2b8528cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DFG = pd.read_csv(output_gene_index)\n",
    "# DFC = pd.read_csv(output_bed_index,sep='\\t',header=None)\n",
    "# G = nx.read_gpickle(output_graph_index)\n",
    "\n",
    "################################\n",
    "########## TODO: ORPHANS ARE NOT A GOOD MEASURE. REPLACE\n",
    "# orphans = [x[0] for x in gs.degree() if x[1]==0]\n",
    "#     gs.remove_nodes_from(set(orphans))    \n",
    "#     nodelist = list(gs.nodes())\n",
    "# WITH A CONNECTIVITY CHECK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5299e4a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "overlap_thr = 0.1\n",
    "#gene_id = 'ENSG00000268388'\n",
    "#gene_id = 'ENSG00000259974'\n",
    "#gene_id = 'ENSG00000280916'\n",
    "#gene_id = 'ENSG00000253827'\n",
    "#gene_id = 'ENSG00000254100'\n",
    "#gene_id = 'ENSG00000251562'\n",
    "#gene_id = 'ENSG00000224397'\n",
    "#gene_id = 'ENSG00000245532'\n",
    "#gene_id = 'ENSG00000245750'\n",
    "gene_id = 'ENSG00000196118'\n",
    "plot_history_network(gene_id,DFG,G,overlap_thr=overlap_thr)\n",
    "plot_history_diagram(gene_id,DFG,G,overlap_thr=overlap_thr)\n",
    "plot_tracks(gene_id,DFG,G,DFC,overlap_thr=overlap_thr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bde24e0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6efcf096",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
